In [1]:
from warnings import filterwarnings
filterwarnings("ignore")

### Step 1 - Data ingestion


In [2]:
import pandas as pd
df = pd.read_csv("training_set.csv", na_values = ["", "NA"], keep_default_na=False)
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

### Target - sales price
### Step 2 - Perfoerm basic data quality checks

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [4]:
m = df.isna().sum()
m

Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

In [5]:
m[m>0]

LotFrontage      259
Alley           1369
MasVnrType         8
MasVnrArea         8
BsmtQual          37
BsmtCond          37
BsmtExposure      38
BsmtFinType1      37
BsmtFinType2      38
Electrical         1
FireplaceQu      690
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
PoolQC          1453
Fence           1179
MiscFeature     1406
dtype: int64

In [6]:
df.duplicated().sum()

np.int64(0)

### step 3 - Seperate X and Y (weight)

In [8]:
X = df.drop(columns=["Id", "SalePrice"])
Y = df[["SalePrice"]]

In [9]:
X.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave   NaN      Reg   
1          20       RL         80.0     9600   Pave   NaN      Reg   
2          60       RL         68.0    11250   Pave   NaN      IR1   
3          70       RL         60.0     9550   Pave   NaN      IR1   
4          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig  ... ScreenPorch PoolArea PoolQC Fence  \
0         Lvl    AllPub    Inside  ...           0        0    NaN   NaN   
1         Lvl    AllPub       FR2  ...           0        0    NaN   NaN   
2         Lvl    AllPub    Inside  ...           0        0    NaN   NaN   
3         Lvl    AllPub    Corner  ...           0        0    NaN   NaN   
4         Lvl    AllPub       FR2  ...           0        0    NaN   NaN   

  MiscFeature MiscVal  MoSold  YrSold  SaleType  SaleCondition  
0         NaN       0       2    2008        WD         Normal  
1         NaN       0       5    2007        WD         Normal  
2         NaN       0       9    2008        WD         Normal  
3         NaN       0       2    2006        WD        Abnorml  
4         NaN       0      12    2008        WD         Normal  

[5 rows x 79 columns]

In [10]:
Y.head()

SalePrice
0     208500
1     181500
2     223500
3     140000
4     250000

### Step 4 - Apply Preprocessing on X

In [11]:
X.dtypes

MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
                  ...   
MiscVal            int64
MoSold             int64
YrSold             int64
SaleType          object
SaleCondition     object
Length: 79, dtype: object

In [12]:
cat = list(X.columns[X.dtypes == "object"])
con = list(X.columns[X.dtypes != "object"])

In [13]:
cat

['MSZoning',
 'Street',
 'Alley',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Neighborhood',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Heating',
 'HeatingQC',
 'CentralAir',
 'Electrical',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageType',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'PoolQC',
 'Fence',
 'MiscFeature',
 'SaleType',
 'SaleCondition']

In [14]:
con

['MSSubClass',
 'LotFrontage',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'YearRemodAdd',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'BsmtHalfBath',
 'FullBath',
 'HalfBath',
 'BedroomAbvGr',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageYrBlt',
 'GarageCars',
 'GarageArea',
 'WoodDeckSF',
 'OpenPorchSF',
 'EnclosedPorch',
 '3SsnPorch',
 'ScreenPorch',
 'PoolArea',
 'MiscVal',
 'MoSold',
 'YrSold']

In [15]:
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer

In [16]:
num_pipe = make_pipeline(
    SimpleImputer(strategy="mean"),
    StandardScaler()
)

In [17]:
cat_pipe = make_pipeline(
    SimpleImputer(strategy="most_frequent"),
    OneHotEncoder(handle_unknown="ignore", sparse_output=False, drop="first")
)

In [18]:
pre = ColumnTransformer(
    [
        ("num", num_pipe, con),
        ("cat", cat_pipe, cat)
    ]
).set_output(transform="pandas")

In [19]:
X_pre = pre.fit_transform(X)
X_pre.head()


num__MSSubClass  num__LotFrontage  num__LotArea  num__OverallQual  \
0         0.073375         -0.229372     -0.207142          0.651479   
1        -0.872563          0.451936     -0.091886         -0.071836   
2         0.073375         -0.093110      0.073480          0.651479   
3         0.309859         -0.456474     -0.096897          0.651479   
4         0.073375          0.633618      0.375148          1.374795   

   num__OverallCond  num__YearBuilt  num__YearRemodAdd  num__MasVnrArea  \
0         -0.517200        1.050994           0.878668         0.511418   
1          2.179628        0.156734          -0.429577        -0.574410   
2         -0.517200        0.984752           0.830215         0.323060   
3         -0.517200       -1.863632          -0.720298        -0.574410   
4         -0.517200        0.951632           0.733308         1.364570   

   num__BsmtFinSF1  num__BsmtFinSF2  ...  cat__SaleType_ConLI  \
0         0.575425        -0.288653  ...                  0.0   
1         1.171992        -0.288653  ...                  0.0   
2         0.092907        -0.288653  ...                  0.0   
3        -0.499274        -0.288653  ...                  0.0   
4         0.463568        -0.288653  ...                  0.0   

   cat__SaleType_ConLw  cat__SaleType_New  cat__SaleType_Oth  \
0                  0.0                0.0                0.0   
1                  0.0                0.0                0.0   
2                  0.0                0.0                0.0   
3                  0.0                0.0                0.0   
4                  0.0                0.0                0.0   

   cat__SaleType_WD  cat__SaleCondition_AdjLand  cat__SaleCondition_Alloca  \
0               1.0                         0.0                        0.0   
1               1.0                         0.0                        0.0   
2               1.0                         0.0                        0.0   
3               1.0                         0.0                        0.0   
4               1.0                         0.0                        0.0   

   cat__SaleCondition_Family  cat__SaleCondition_Normal  \
0                        0.0                        1.0   
1                        0.0                        1.0   
2                        0.0                        1.0   
3                        0.0                        0.0   
4                        0.0                        1.0   

   cat__SaleCondition_Partial  
0                         0.0  
1                         0.0  
2                         0.0  
3                         0.0  
4                         0.0  

[5 rows x 245 columns]

### Step 5 - Apply Train test split

In [20]:
from sklearn.model_selection import train_test_split

xtrain, xtest, ytrain, ytest = train_test_split(X_pre, Y, test_size=0.2, random_state=10)

In [21]:
xtrain.head()

num__MSSubClass  num__LotFrontage  num__LotArea  num__OverallQual  \
1216         0.782828     -9.311018e-02     -0.159035         -0.071836   
339         -0.872563     -1.839512e-01      0.188735         -0.071836   
1057         0.073375      6.454645e-16      1.948534          0.651479   
482          0.309859     -9.106796e-01     -0.803463          0.651479   
529         -0.872563      6.454645e-16      2.220035         -0.071836   

      num__OverallCond  num__YearBuilt  num__YearRemodAdd  num__MasVnrArea  \
1216         -0.517200        0.222975          -0.332670         -0.57441   
339           1.280685       -0.439440           0.636400          0.40062   
1057          0.381743        0.752907           0.442586         -0.57441   
482           2.179628       -1.863632           0.975575         -0.57441   
529          -2.315085       -0.472560          -0.478030          0.00000   

      num__BsmtFinSF1  num__BsmtFinSF2  ...  cat__SaleType_ConLI  \
1216        -0.973018        -0.288653  ...                  0.0   
339          0.310040        -0.288653  ...                  0.0   
1057         0.331973        -0.288653  ...                  0.0   
482         -0.317233        -0.288653  ...                  0.0   
529          1.700568        -0.288653  ...                  0.0   

      cat__SaleType_ConLw  cat__SaleType_New  cat__SaleType_Oth  \
1216                  0.0                0.0                0.0   
339                   0.0                0.0                0.0   
1057                  0.0                0.0                0.0   
482                   0.0                0.0                0.0   
529                   0.0                0.0                0.0   

      cat__SaleType_WD  cat__SaleCondition_AdjLand  cat__SaleCondition_Alloca  \
1216               1.0                         0.0                        0.0   
339                1.0                         0.0                        0.0   
1057               1.0                         0.0                        0.0   
482                1.0                         0.0                        0.0   
529                1.0                         0.0                        1.0   

      cat__SaleCondition_Family  cat__SaleCondition_Normal  \
1216                        0.0                        1.0   
339                         0.0                        1.0   
1057                        0.0                        1.0   
482                         0.0                        1.0   
529                         0.0                        0.0   

      cat__SaleCondition_Partial  
1216                         0.0  
339                          0.0  
1057                         0.0  
482                          0.0  
529                          0.0  

[5 rows x 245 columns]

In [22]:

ytrain.head()

SalePrice
1216     112000
339      155000
1057     248000
482      155000
529      200624

In [23]:
xtest.head()

num__MSSubClass  num__LotFrontage  num__LotArea  num__OverallQual  \
854        -0.872563      1.451188e+00      0.741961         -0.795151   
381        -0.872563     -4.564744e-01     -0.332419          0.651479   
816        -0.872563      6.454645e-16      0.091019         -0.795151   
577         0.546344      1.178664e+00      0.126297         -0.795151   
35          0.073375      1.723711e+00      0.290761          1.374795   

     num__OverallCond  num__YearBuilt  num__YearRemodAdd  num__MasVnrArea  \
854         -1.416142       -0.538802          -0.526484        -0.574410   
381         -0.517200        1.150356           1.024029        -0.574410   
816          0.381743       -0.571923          -1.495554        -0.574410   
577          0.381743       -0.174474          -0.914112        -0.037036   
35          -0.517200        1.084115           0.975575         0.156862   

     num__BsmtFinSF1  num__BsmtFinSF2  ...  cat__SaleType_ConLI  \
854        -0.301880         6.439444  ...                  0.0   
381        -0.973018        -0.288653  ...                  0.0   
816         0.092907        -0.288653  ...                  0.0   
577        -0.253628         3.128104  ...                  0.0   
35         -0.973018        -0.288653  ...                  0.0   

     cat__SaleType_ConLw  cat__SaleType_New  cat__SaleType_Oth  \
854                  0.0                0.0                0.0   
381                  0.0                1.0                0.0   
816                  0.0                0.0                0.0   
577                  0.0                0.0                0.0   
35                   0.0                0.0                0.0   

     cat__SaleType_WD  cat__SaleCondition_AdjLand  cat__SaleCondition_Alloca  \
854               1.0                         0.0                        0.0   
381               0.0                         0.0                        0.0   
816               1.0                         0.0                        0.0   
577               1.0                         0.0                        0.0   
35                1.0                         0.0                        0.0   

     cat__SaleCondition_Family  cat__SaleCondition_Normal  \
854                        0.0                        0.0   
381                        0.0                        0.0   
816                        0.0                        1.0   
577                        0.0                        0.0   
35                         0.0                        1.0   

     cat__SaleCondition_Partial  
854                         0.0  
381                         1.0  
816                         0.0  
577                         0.0  
35                          0.0  

[5 rows x 245 columns]

In [24]:
ytest.head()


SalePrice
854     170000
381     187750
816     137000
577     164500
35      309000

In [25]:
xtrain.shape

(1168, 245)

In [26]:
xtest.shape

(292, 245)

### Step 6 - Model Building

In [27]:
from sklearn.linear_model import LinearRegression
model1 = LinearRegression()
model1.fit(xtrain, ytrain)

LinearRegression()

In [28]:
model1.score(xtrain, ytrain)

0.9343927883411204

In [29]:

model1.score(xtest, ytest)

0.4129817529082652

### ridge Model
Penalty applied on square of coefficents

alpha is hypermeter

optimum value of alpha is unknown

In [30]:
from sklearn.linear_model import Ridge
model_ridge = Ridge(alpha=1)
model_ridge.fit(xtrain, ytrain)

Ridge(alpha=1)

In [31]:
model_ridge.score(xtrain, ytrain)

0.9113005164468516

In [32]:
model_ridge.score(xtest, ytest)

0.8872780274844718

In [33]:
from sklearn.model_selection import cross_val_score

scores1 = cross_val_score(model_ridge, xtrain, ytrain, cv=5, scoring="r2", n_jobs=-1)

scores1

array([0.85003995, 0.86869834, 0.57939032, 0.73883037, 0.83611872])

In [34]:
scores1.mean()

np.float64(0.774615537598111)

### Model Tuning
Optimizing value

GridSearchCV




In [35]:
import numpy as np
params1 = {"alpha" : np.arange(start=1, stop=100, step=0.5)}
params1

{'alpha': array([ 1. ,  1.5,  2. ,  2.5,  3. ,  3.5,  4. ,  4.5,  5. ,  5.5,  6. ,
         6.5,  7. ,  7.5,  8. ,  8.5,  9. ,  9.5, 10. , 10.5, 11. , 11.5,
        12. , 12.5, 13. , 13.5, 14. , 14.5, 15. , 15.5, 16. , 16.5, 17. ,
        17.5, 18. , 18.5, 19. , 19.5, 20. , 20.5, 21. , 21.5, 22. , 22.5,
        23. , 23.5, 24. , 24.5, 25. , 25.5, 26. , 26.5, 27. , 27.5, 28. ,
        28.5, 29. , 29.5, 30. , 30.5, 31. , 31.5, 32. , 32.5, 33. , 33.5,
        34. , 34.5, 35. , 35.5, 36. , 36.5, 37. , 37.5, 38. , 38.5, 39. ,
        39.5, 40. , 40.5, 41. , 41.5, 42. , 42.5, 43. , 43.5, 44. , 44.5,
        45. , 45.5, 46. , 46.5, 47. , 47.5, 48. , 48.5, 49. , 49.5, 50. ,
        50.5, 51. , 51.5, 52. , 52.5, 53. , 53.5, 54. , 54.5, 55. , 55.5,
        56. , 56.5, 57. , 57.5, 58. , 58.5, 59. , 59.5, 60. , 60.5, 61. ,
        61.5, 62. , 62.5, 63. , 63.5, 64. , 64.5, 65. , 65.5, 66. , 66.5,
        67. , 67.5, 68. , 68.5, 69. , 69.5, 70. , 70.5, 71. , 71.5, 72. ,
        72.5, 73. , 73.5, 74.

In [37]:
from sklearn.model_selection import GridSearchCV

In [38]:
base_ridge = Ridge()
gscv_ridge = GridSearchCV(base_ridge, param_grid=params1, cv=5, scoring="r2", n_jobs=-1)
gscv_ridge.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Ridge(), n_jobs=-1,
             param_grid={'alpha': array([ 1. ,  1.5,  2. ,  2.5,  3. ,  3.5,  4. ,  4.5,  5. ,  5.5,  6. ,
        6.5,  7. ,  7.5,  8. ,  8.5,  9. ,  9.5, 10. , 10.5, 11. , 11.5,
       12. , 12.5, 13. , 13.5, 14. , 14.5, 15. , 15.5, 16. , 16.5, 17. ,
       17.5, 18. , 18.5, 19. , 19.5, 20. , 20.5, 21. , 21.5, 22. , 22.5,
       23. , 23.5, 24. , 24.5, 25. , 25.5, 26. , 26.5, 27. , 27.5, 28. ,
       28.5, 29. , 29.5, 30. , 30.5, 31. , 31.5, 32. , 32.5, 33. , 33....
       61.5, 62. , 62.5, 63. , 63.5, 64. , 64.5, 65. , 65.5, 66. , 66.5,
       67. , 67.5, 68. , 68.5, 69. , 69.5, 70. , 70.5, 71. , 71.5, 72. ,
       72.5, 73. , 73.5, 74. , 74.5, 75. , 75.5, 76. , 76.5, 77. , 77.5,
       78. , 78.5, 79. , 79.5, 80. , 80.5, 81. , 81.5, 82. , 82.5, 83. ,
       83.5, 84. , 84.5, 85. , 85.5, 86. , 86.5, 87. , 87.5, 88. , 88.5,
       89. , 89.5, 90. , 90.5, 91. , 91.5, 92. , 92.5, 93. , 93.5, 94. ,
       94.5, 95. , 95.5, 96. , 96.5, 97. , 97.5, 98. , 98.5, 99. , 99.5])},
             scoring='r2')

In [39]:

gscv_ridge.best_params_

{'alpha': np.float64(16.0)}

In [40]:
gscv_ridge.best_score_

np.float64(0.7920183087955251)

In [41]:
best_ridge = gscv_ridge.best_estimator_
best_ridge

Ridge(alpha=np.float64(16.0))

In [42]:
best_ridge.score(xtrain, ytrain)

0.8866827780575954

In [43]:

best_ridge.score(xtest, ytest)

0.8918637590610943

### 2.Lasso
Penalty applied on absolute value of coefficient

In [44]:
from sklearn.linear_model import Lasso
model_lasso = Lasso(alpha=1)
model_lasso.fit(xtrain, ytrain)

Lasso(alpha=1)

In [45]:
model_lasso.score(xtrain, ytrain)

0.9342970028602909

In [46]:

model_lasso.score(xtest, ytest)

0.44479027175280283

### Hyperparameter tuning Lasso

In [47]:
params1

{'alpha': array([ 1. ,  1.5,  2. ,  2.5,  3. ,  3.5,  4. ,  4.5,  5. ,  5.5,  6. ,
         6.5,  7. ,  7.5,  8. ,  8.5,  9. ,  9.5, 10. , 10.5, 11. , 11.5,
        12. , 12.5, 13. , 13.5, 14. , 14.5, 15. , 15.5, 16. , 16.5, 17. ,
        17.5, 18. , 18.5, 19. , 19.5, 20. , 20.5, 21. , 21.5, 22. , 22.5,
        23. , 23.5, 24. , 24.5, 25. , 25.5, 26. , 26.5, 27. , 27.5, 28. ,
        28.5, 29. , 29.5, 30. , 30.5, 31. , 31.5, 32. , 32.5, 33. , 33.5,
        34. , 34.5, 35. , 35.5, 36. , 36.5, 37. , 37.5, 38. , 38.5, 39. ,
        39.5, 40. , 40.5, 41. , 41.5, 42. , 42.5, 43. , 43.5, 44. , 44.5,
        45. , 45.5, 46. , 46.5, 47. , 47.5, 48. , 48.5, 49. , 49.5, 50. ,
        50.5, 51. , 51.5, 52. , 52.5, 53. , 53.5, 54. , 54.5, 55. , 55.5,
        56. , 56.5, 57. , 57.5, 58. , 58.5, 59. , 59.5, 60. , 60.5, 61. ,
        61.5, 62. , 62.5, 63. , 63.5, 64. , 64.5, 65. , 65.5, 66. , 66.5,
        67. , 67.5, 68. , 68.5, 69. , 69.5, 70. , 70.5, 71. , 71.5, 72. ,
        72.5, 73. , 73.5, 74.

In [48]:
base_lasso = Lasso()
gscv_lasso = GridSearchCV(base_lasso, param_grid=params1, cv=5, scoring="r2", n_jobs=-1)
gscv_lasso.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Lasso(), n_jobs=-1,
             param_grid={'alpha': array([ 1. ,  1.5,  2. ,  2.5,  3. ,  3.5,  4. ,  4.5,  5. ,  5.5,  6. ,
        6.5,  7. ,  7.5,  8. ,  8.5,  9. ,  9.5, 10. , 10.5, 11. , 11.5,
       12. , 12.5, 13. , 13.5, 14. , 14.5, 15. , 15.5, 16. , 16.5, 17. ,
       17.5, 18. , 18.5, 19. , 19.5, 20. , 20.5, 21. , 21.5, 22. , 22.5,
       23. , 23.5, 24. , 24.5, 25. , 25.5, 26. , 26.5, 27. , 27.5, 28. ,
       28.5, 29. , 29.5, 30. , 30.5, 31. , 31.5, 32. , 32.5, 33. , 33....
       61.5, 62. , 62.5, 63. , 63.5, 64. , 64.5, 65. , 65.5, 66. , 66.5,
       67. , 67.5, 68. , 68.5, 69. , 69.5, 70. , 70.5, 71. , 71.5, 72. ,
       72.5, 73. , 73.5, 74. , 74.5, 75. , 75.5, 76. , 76.5, 77. , 77.5,
       78. , 78.5, 79. , 79.5, 80. , 80.5, 81. , 81.5, 82. , 82.5, 83. ,
       83.5, 84. , 84.5, 85. , 85.5, 86. , 86.5, 87. , 87.5, 88. , 88.5,
       89. , 89.5, 90. , 90.5, 91. , 91.5, 92. , 92.5, 93. , 93.5, 94. ,
       94.5, 95. , 95.5, 96. , 96.5, 97. , 97.5, 98. , 98.5, 99. , 99.5])},
             scoring='r2')

In [49]:
gscv_lasso.best_params_

{'alpha': np.float64(99.5)}

In [50]:
gscv_lasso.best_score_

np.float64(0.7889407135548316)

In [51]:
best_lasso = gscv_lasso.best_estimator_
best_lasso

Lasso(alpha=np.float64(99.5))

In [52]:

best_lasso.score(xtrain, ytrain)

0.8972279712871012

In [53]:
best_lasso.score(xtest, ytest)

0.8951178080672308

### From above result lasso model has best r2 score in best which is 0.8951

### Step 7 - Evaluate  lasso model in detail 

In [54]:
from sklearn.metrics import (
    root_mean_squared_error, mean_absolute_error, mean_absolute_percentage_error, r2_score
)

In [55]:
def evaluate_model(model, x, y):
    ypred = model.predict(x)
    rmse = root_mean_squared_error(y, ypred)
    mae = mean_absolute_error(y, ypred)
    mape = mean_absolute_percentage_error(y, ypred)
    r2 = r2_score(y, ypred)
    print(f"RMSE : {rmse:.2f}")
    print(f"MAE : {mae:.2f}")
    print(f"MAPE : {mape:.2%}")
    print(f"R2 : {r2:.2%}")

In [56]:
print("Lasso Model Train :")
evaluate_model(best_ridge, xtrain, ytrain)

Lasso Model Train :
RMSE : 26682.84
MAE : 15616.25
MAPE : 9.08%
R2 : 88.67%


In [57]:
print("Lasso Model Test:")
evaluate_model(best_ridge, xtest, ytest)

Lasso Model Test:
RMSE : 26301.05
MAE : 18558.99
MAPE : 11.08%
R2 : 89.19%


### Step 8 - Out of sample prediction

In [58]:
xnew = pd.read_csv("sample_set.csv", na_values=["", "NA"], keep_default_na=False)
xnew

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence  \
0            Lvl    AllPub  ...         120        0    NaN  MnPrv   
1            Lvl    AllPub  ...           0        0    NaN    NaN   
2            Lvl    AllPub  ...           0        0    NaN  MnPrv   
3            Lvl    AllPub  ...           0        0    NaN    NaN   
4            HLS    AllPub  ...         144        0    NaN    NaN   
...          ...       ...  ...         ...      ...    ...    ...   
1454         Lvl    AllPub  ...           0        0    NaN    NaN   
1455         Lvl    AllPub  ...           0        0    NaN    NaN   
1456         Lvl    AllPub  ...           0        0    NaN    NaN   
1457         Lvl    AllPub  ...           0        0    NaN  MnPrv   
1458         Lvl    AllPub  ...           0        0    NaN    NaN   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0            NaN       0      6    2010        WD         Normal  
1           Gar2   12500      6    2010        WD         Normal  
2            NaN       0      3    2010        WD         Normal  
3            NaN       0      6    2010        WD         Normal  
4            NaN       0      1    2010        WD         Normal  
...          ...     ...    ...     ...       ...            ...  
1454         NaN       0      6    2006        WD         Normal  
1455         NaN       0      4    2006        WD        Abnorml  
1456         NaN       0      9    2006        WD        Abnorml  
1457        Shed     700      7    2006        WD         Normal  
1458         NaN       0     11    2006        WD         Normal  

[1459 rows x 80 columns]

In [59]:
pre

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 ['MSSubClass', 'LotFrontage', 'LotArea',
                                  'OverallQual', 'OverallCond', 'YearBuilt',
                                  'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
                                  'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF',
                                  '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
                                  'GrLivArea', 'BsmtFullBath', '...
                                 ['MSZoning', 'Street', 'Alley', 'LotShape',
                                  'LandContour', 'Utilities', 'LotConfig',
                                  'LandSlope', 'Neighborhood', 'Condition1',
                                  'Condition2', 'BldgType', 'HouseStyle',
                                  'RoofStyle', 'RoofMatl', 'Exterior1st',
                                  'Exterior2nd', 'MasVnrType', 'ExterQual',
                                  'ExterCond', 'Foundation', 'BsmtQual',
                                  'BsmtCond', 'BsmtExposure', 'BsmtFinType1',
                                  'BsmtFinType2', 'Heating', 'HeatingQC',
                                  'CentralAir', 'Electrical', ...])])

In [60]:
xnew_pre = pre.transform(xnew)
xnew_pre

num__MSSubClass  num__LotFrontage  num__LotArea  num__OverallQual  \
0           -0.872563          0.451936      0.110763         -0.795151   
1           -0.872563          0.497357      0.375850         -0.071836   
2            0.073375          0.179413      0.332053         -0.795151   
3            0.073375          0.361095     -0.054002         -0.071836   
4            1.492282         -1.228623     -0.552407          1.374795   
...               ...               ...           ...               ...   
1454         2.438219         -2.227875     -0.859988         -1.518467   
1455         2.438219         -2.227875     -0.864197         -1.518467   
1456        -0.872563          4.085578      0.950423         -0.795151   
1457         0.664586         -0.365633     -0.007600         -0.795151   
1458         0.073375          0.179413     -0.089180          0.651479   

      num__OverallCond  num__YearBuilt  num__YearRemodAdd  num__MasVnrArea  \
0             0.381743       -0.340077          -1.156380        -0.574410   
1             0.381743       -0.439440          -1.301740         0.023903   
2            -0.517200        0.852269           0.636400        -0.574410   
3             0.381743        0.885390           0.636400        -0.463612   
4            -0.517200        0.686666           0.345679        -0.574410   
...                ...             ...                ...              ...   
1454          1.280685       -0.041991          -0.720298        -0.574410   
1455         -0.517200       -0.041991          -0.720298        -0.574410   
1456          1.280685       -0.373198           0.539493        -0.574410   
1457         -0.517200        0.686666           0.345679        -0.574410   
1458         -0.517200        0.719786           0.442586        -0.053656   

      num__BsmtFinSF1  num__BsmtFinSF2  ...  cat__SaleType_ConLI  \
0            0.053428         0.604293  ...                  0.0   
1            1.051363        -0.288653  ...                  0.0   
2            0.761852        -0.288653  ...                  0.0   
3            0.347326        -0.288653  ...                  0.0   
4           -0.396190        -0.288653  ...                  0.0   
...               ...              ...  ...                  ...   
1454        -0.973018        -0.288653  ...                  0.0   
1455        -0.420316        -0.288653  ...                  0.0   
1456         1.711535        -0.288653  ...                  0.0   
1457        -0.233889        -0.288653  ...                  0.0   
1458         0.689474        -0.288653  ...                  0.0   

      cat__SaleType_ConLw  cat__SaleType_New  cat__SaleType_Oth  \
0                     0.0                0.0                0.0   
1                     0.0                0.0                0.0   
2                     0.0                0.0                0.0   
3                     0.0                0.0                0.0   
4                     0.0                0.0                0.0   
...                   ...                ...                ...   
1454                  0.0                0.0                0.0   
1455                  0.0                0.0                0.0   
1456                  0.0                0.0                0.0   
1457                  0.0                0.0                0.0   
1458                  0.0                0.0                0.0   

      cat__SaleType_WD  cat__SaleCondition_AdjLand  cat__SaleCondition_Alloca  \
0                  1.0                         0.0                        0.0   
1                  1.0                         0.0                        0.0   
2                  1.0                         0.0                        0.0   
3                  1.0                         0.0                        0.0   
4                  1.0                         0.0                        0.0   
...                ...                         ...          

### Selecting best ridge model because of highest r2 score

In [61]:
preds = best_lasso.predict(xnew_pre)
preds

array([107525.78498249, 138683.9058278 , 174239.27624958, ...,
       157094.80250916, 105482.29206354, 236167.81655203], shape=(1459,))

In [62]:
xnew["SalePrice_pred"] = preds

In [63]:
xnew = xnew.round(2)
xnew

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN        Gar2   12500   
2            Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            HLS    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1454         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  MnPrv        Shed     700   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice_pred  
0         6   2010        WD         Normal       107525.78  
1         6   2010        WD         Normal       138683.91  
2         3   2010        WD         Normal       174239.28  
3         6   2010        WD         Normal       191342.26  
4         1   2010        WD         Normal       206757.44  
...     ...    ...       ...            ...             ...  
1454      6   2006        WD         Normal        77699.27  
1455      4   2006        WD        Abnorml        79614.19  
1456      9   2006        WD        Abnorml       157094.80  
1457      7   2006        WD         Normal       105482.29  
1458     11   2006        WD         Normal       236167.82  

[1459 rows x 81 columns]

In [64]:
xnew.to_csv("LassoResults.csv", index=False)

### Save best model and preprocesssor in joblib format

In [65]:
pre

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 ['MSSubClass', 'LotFrontage', 'LotArea',
                                  'OverallQual', 'OverallCond', 'YearBuilt',
                                  'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
                                  'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF',
                                  '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
                                  'GrLivArea', 'BsmtFullBath', '...
                                 ['MSZoning', 'Street', 'Alley', 'LotShape',
                                  'LandContour', 'Utilities', 'LotConfig',
                                  'LandSlope', 'Neighborhood', 'Condition1',
                                  'Condition2', 'BldgType', 'HouseStyle',
                                  'RoofStyle', 'RoofMatl', 'Exterior1st',
                                  'Exterior2nd', 'MasVnrType', 'ExterQual',
                                  'ExterCond', 'Foundation', 'BsmtQual',
                                  'BsmtCond', 'BsmtExposure', 'BsmtFinType1',
                                  'BsmtFinType2', 'Heating', 'HeatingQC',
                                  'CentralAir', 'Electrical', ...])])

In [66]:
best_lasso

Lasso(alpha=np.float64(99.5))

In [67]:

%pip install joblib

Note: you may need to restart the kernel to use updated packages.


In [68]:
import joblib

joblib.dump(pre, "pre.joblib")

['pre.joblib']

In [69]:
joblib.dump(best_lasso, "lasso_model.joblib")

['lasso_model.joblib']

### Loading pre and model using joblib

In [70]:
p = joblib.load("pre.joblib")
p

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 ['MSSubClass', 'LotFrontage', 'LotArea',
                                  'OverallQual', 'OverallCond', 'YearBuilt',
                                  'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
                                  'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF',
                                  '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
                                  'GrLivArea', 'BsmtFullBath', '...
                                 ['MSZoning', 'Street', 'Alley', 'LotShape',
                                  'LandContour', 'Utilities', 'LotConfig',
                                  'LandSlope', 'Neighborhood', 'Condition1',
                                  'Condition2', 'BldgType', 'HouseStyle',
                                  'RoofStyle', 'RoofMatl', 'Exterior1st',
                                  'Exterior2nd', 'MasVnrType', 'ExterQual',
                                  'ExterCond', 'Foundation', 'BsmtQual',
                                  'BsmtCond', 'BsmtExposure', 'BsmtFinType1',
                                  'BsmtFinType2', 'Heating', 'HeatingQC',
                                  'CentralAir', 'Electrical', ...])])

In [72]:
m = joblib.load("lasso_model.joblib")
m

Lasso(alpha=np.float64(99.5))